cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -367,12 +379,106 @@ class device_row_comparator {
+      column_device_view rcol = _rhs.slice(rhs_element_index, 1);
+      while (lcol.type().id() == type_id::LIST) {
+        lcol = detail::lists_column_device_view(lcol).get_sliced_child();
+        rcol = detail::lists_column_device_view(rcol).get_sliced_child();
+      }
+
+      // These start and end values indicate the start and end points of all
+      // the elements of the lists in the current list element
+      // (`[lhs|rhs]_element_index`) that we are comparing.
+      auto const l_offsets = _l_dremel_device_view.offsets;
+      auto const r_offsets = _r_dremel_device_view.offsets;
+      auto l_start         = l_offsets[lhs_element_index];
+      auto l_end           = l_offsets[lhs_element_index + 1];
+      auto r_start         = r_offsets[rhs_element_index];
+      auto r_end           = r_offsets[rhs_element_index + 1];
+
+      // This comparator will be used when we actually need to compare elements
+      auto comparator =
+        element_comparator{_check_nulls, lcol, rcol, _null_precedence, _depth, _comparator};
+
+      // Loop over each element in the encoding. Note that this includes nulls
+      // and empty lists, so not every index i/j correspondings to an actual
+      // element in the child column. The index k is used to keep track of the
+      // current child element that we're actually comparing.
+      weak_ordering state{weak_ordering::EQUIVALENT};
+      for (int left_dremel_index = l_start, right_dremel_index = r_start, element_index = 0;
+           left_dremel_index < l_end and right_dremel_index < r_end;
+           ++left_dremel_index, ++right_dremel_index) {
+        // First early exit: the definition levels do not match.
+        if (l_def_levels[left_dremel_index] != r_def_levels[right_dremel_index]) {
+          state = (l_def_levels[left_dremel_index] < r_def_levels[right_dremel_index])
+                    ? weak_ordering::LESS
+                    : weak_ordering::GREATER;
+          return cuda::std::pair(state, _depth);
+        }
+
+        // Second early exit: the repetition levels do not match.
+        if (l_rep_levels[left_dremel_index] != r_rep_levels[right_dremel_index]) {
+          state = (l_rep_levels[left_dremel_index] < r_rep_levels[right_dremel_index])
+                    ? weak_ordering::LESS
+                    : weak_ordering::GREATER;
+          return cuda::std::pair(state, _depth);
+        }
+
+        // Third early exit: This case has two branches.
+        // 1) If we are at the maximum definition level, then we actually have
+        //    an underlying element to compare, not just an empty list or a
+        //    null. Therefore, we access the kth element of each list and
+        //    compare the values.
+        // 2) If we are 1 - the maximum definition level and the column is
+        //    nullable, we know that we are looking at a element in the child
+        //    column. In this case we simply skip to the next element.
```

Comment:
This second half of the comment is so confusing... I'm trying to come up with a rephrasing but I've been spinning my wheels for a bit. Giving up for now and will revisit this thread later.

Reply:
I've rewritten this. There's a key word "null" missing on the second line of step 2), which I have now added while also rewording the rest of it. Hopefully it's clear, now.